File: app/src/main/java/kandroid/proxy/MyProxyServlet.java
package kandroid.proxy;

import org.apache.commons.io.output.ByteArrayOutputStream;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.HttpProxy;
import org.eclipse.jetty.client.ProxyConfiguration;
import org.eclipse.jetty.client.api.Request;
import org.eclipse.jetty.client.api.Response;
import org.eclipse.jetty.proxy.AsyncProxyServlet;
import org.eclipse.jetty.util.Callback;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import kandroid.config.Config;
import kandroid.observer.RawData;
import kandroid.observer.ApiLoader;

public class MyProxyServlet extends AsyncProxyServlet {

    @Override
    protected void sendProxyRequest(HttpServletRequest clientRequest, HttpServletResponse proxyResponse,
                                    Request proxyRequest) {
        proxyRequest.onRequestContent(new RequestContentListener(clientRequest));
        super.sendProxyRequest(clientRequest, proxyResponse, proxyRequest);
    }

    @Override
    protected void onResponseContent(HttpServletRequest request, HttpServletResponse response, Response proxyResponse,
                                     byte[] buffer, int offset, int length, Callback callback) {
        if (MyFilter.isNeed(request.getServerName(), response.getContentType())) {
            ByteArrayOutputStream stream = (ByteArrayOutputStream) request.getAttribute(MyFilter.RESPONSE_BODY);
            if (stream == null) {
                stream = new ByteArrayOutputStream();
                request.setAttribute(MyFilter.RESPONSE_BODY, stream);
            }
            stream.write(buffer, offset, length);
        }
        super.onResponseContent(request, response, proxyResponse, buffer, offset, length, callback);
    }

    @Override
    protected void onProxyResponseSuccess(HttpServletRequest request, HttpServletResponse response,
                                          Response proxyResponse) {
        System.out.println(request.getServerName() + " | " + request.getRequestURI());
        if (MyFilter.isNeed(request.getServerName(), response.getContentType())) {
            ByteArrayOutputStream stream = (ByteArrayOutputStream) request.getAttribute(MyFilter.RESPONSE_BODY);
            if (stream != null) {
                byte[] postField = (byte[]) request.getAttribute(MyFilter.REQUEST_BODY);
                try {
                    System.out.println(URLDecoder.decode(new String(postField),"UTF-8"));
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
                RawData data = new RawData(request.getRequestURI(), postField, stream.toByteArray());
                ApiLoader.load(data);
            }
        }
        super.onProxyResponseSuccess(request, response, proxyResponse);
    }

    @Override
    protected HttpClient newHttpClient() {
        HttpClient client = super.newHttpClient();
        if (Config.get().isUseProxy()) {
            String host = Config.get().getProxyHost();
            int port = Config.get().getProxyPort();
            List<ProxyConfiguration.Proxy> proxies = client.getProxyConfiguration().getProxies();
            proxies.add(new HttpProxy(host, port));
        }
        return client;
    }

    private final class RequestContentListener implements Request.ContentListener {

        private final HttpServletRequest httpRequest;

        RequestContentListener(HttpServletRequest request) {
            this.httpRequest = request;
        }

        @Override
        public void onContent(Request request, ByteBuffer buffer) {
            if (((buffer.limit() > 0) && (buffer.limit() <= MyFilter.MAX_POST_FIELD_SIZE))
                    && MyFilter.filterServerName(request.getHost())) {
                this.httpRequest.setAttribute(MyFilter.REQUEST_BODY, Arrays.copyOf(buffer.array(), buffer.limit()));
            }
        }
    }
}

File: app/src/main/java/kandroid/proxy/MyProxyServlet.kt
package kandroid.proxy

import kandroid.config.Config
import kandroid.observer.ApiLoader
import kandroid.observer.RawData
import org.apache.commons.io.output.ByteArrayOutputStream
import org.eclipse.jetty.client.HttpClient
import org.eclipse.jetty.client.HttpProxy
import org.eclipse.jetty.client.api.Request
import org.eclipse.jetty.client.api.Response
import org.eclipse.jetty.proxy.AsyncProxyServlet
import org.eclipse.jetty.util.Callback
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.nio.ByteBuffer
import java.util.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class MyProxyServlet : AsyncProxyServlet() {

    override fun sendProxyRequest(clientRequest: HttpServletRequest, proxyResponse: HttpServletResponse,
                                  proxyRequest: Request) {
        proxyRequest.onRequestContent(RequestContentListener(clientRequest))
        super.sendProxyRequest(clientRequest, proxyResponse, proxyRequest)
    }

    override fun onResponseContent(request: HttpServletRequest, response: HttpServletResponse, proxyResponse: Response?,
                                   buffer: ByteArray, offset: Int, length: Int, callback: Callback) {
        if (MyFilter.isNeed(request.serverName, response.contentType)) {
            var stream = request.getAttribute(MyFilter.RESPONSE_BODY) as ByteArrayOutputStream?
            if (stream == null) {
                stream = ByteArrayOutputStream()
                request.setAttribute(MyFilter.RESPONSE_BODY, stream)
            }
            stream.write(buffer, offset, length)
        }
        super.onResponseContent(request, response, proxyResponse, buffer, offset, length, callback)
    }

    override fun onProxyResponseSuccess(request: HttpServletRequest, response: HttpServletResponse,
                                        proxyResponse: Response?) {
        println(request.serverName + " | " + request.requestURI)
        if (MyFilter.isNeed(request.serverName, response.contentType)) {
            val stream = request.getAttribute(MyFilter.RESPONSE_BODY) as ByteArrayOutputStream?
            if (stream != null) {
                val postField = request.getAttribute(MyFilter.REQUEST_BODY) as ByteArray
                try {
                    println(URLDecoder.decode(String(postField), "UTF-8"))
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                }

                val data = RawData(request.requestURI, postField, stream.toByteArray())
                ApiLoader.load(data)
            }
        }
        super.onProxyResponseSuccess(request, response, proxyResponse)
    }

    override fun newHttpClient(): HttpClient {
        val client = super.newHttpClient()
        if (Config.config.isUseProxy) {
            val host = Config.config.proxyHost
            val port = Config.config.proxyPort
            val proxies = client.proxyConfiguration.proxies
            proxies.add(HttpProxy(host, port))
        }
        return client
    }

    private inner class RequestContentListener(val httpRequest: HttpServletRequest) : Request.ContentListener {

        override fun onContent(request: Request, buffer: ByteBuffer) {
            if (buffer.limit() in 1..MyFilter.MAX_POST_FIELD_SIZE && MyFilter.filterServerName(request.host)) {
                httpRequest.setAttribute(MyFilter.REQUEST_BODY, Arrays.copyOf(buffer.array(), buffer.limit()))
            }
        }
    }
}
